src/sources/exec.rs

```diff
@@ -0,0 +1,830 @@
+                run_scheduled(self.clone(), hostname, exec_interval_secs, shutdown, out)
+            }
+            Mode::Streaming {
+                respawn_on_exit,
+                respawn_interval_secs,
+            } => run_streaming(
+                self.clone(),
+                hostname,
+                respawn_on_exit,
+                respawn_interval_secs,
+                shutdown,
+                out,
+            ),
+        }
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        EXEC
+    }
+}
+
+pub fn run_scheduled(
+    config: ExecConfig,
+    hostname: Option<String>,
+    exec_interval_secs: u64,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> crate::Result<super::Source> {
+    Ok(Box::pin(async move {
+        info!("Starting scheduled exec runs.");
+        let schedule = Duration::from_secs(exec_interval_secs);
+
+        let mut interval =
+            IntervalStream::new(time::interval(schedule)).take_until(shutdown.clone());
+
+        while interval.next().await.is_some() {
+            // Mark the start time just before spawning the process as
+            // this seems to be the best approximation of exec duration
+            let now = Instant::now();
+
+            // Wait for our task to finish, wrapping it in a timeout
+            let timeout = tokio::time::timeout(
+                schedule,
+                run_command(
+                    config.clone(),
+                    hostname.clone(),
+                    shutdown.clone(),
+                    out.clone(),
+                ),
+            );
+
+            let timeout_result = timeout.await;
+
+            match timeout_result {
+                Ok(output) => {
+                    if let Err(command_error) = output {
+                        emit!(ExecFailed {
+                            command: config.command_line().as_str(),
+                            error: command_error,
+                        });
+                    }
+                }
+                Err(_) => {
+                    emit!(ExecTimeout {
+                        command: config.command_line().as_str(),
+                        elapsed_seconds: now.elapsed().as_secs(),
+                    });
+                }
+            }
+        }
+
+        info!("Finished scheduled exec runs.");
+        Ok(())
+    }))
+}
+
+pub fn run_streaming(
+    config: ExecConfig,
+    hostname: Option<String>,
+    respawn_on_exit: bool,
+    respawn_interval_secs: u64,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> crate::Result<super::Source> {
+    Ok(Box::pin(async move {
+        if respawn_on_exit {
+            let duration = Duration::from_secs(respawn_interval_secs);
+
+            // Continue to loop while not shutdown
+            loop {
+                tokio::select! {
+                    _ = shutdown.clone() => break, // will break early if a shutdown is started
```

Comment:
Even with `mut shutdown` I have been unable to get `&mut shutdown` to work in this method because of the combination of mutable and immutable borrows and the loop.

Reply:
Ah I see 🤔